fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/dao/impl/FetchResourceChangesDAO.java

```diff
@@ -0,0 +1,134 @@
+     * @param resourceTypeName
+     * @param fromTstamp
+     * @param afterResourceId
+     */
+    public FetchResourceChangesDAO(IDatabaseTranslator tx, String schemaName, int resourceCount, String resourceTypeName, Instant fromTstamp, Long afterResourceId) {
+        this.translator = tx;
+        this.schemaName = schemaName;
+        this.resourceCount = resourceCount;
+        this.resourceTypeName = resourceTypeName;
+        this.fromTstamp = fromTstamp;
+        this.afterResourceId = afterResourceId;
+    }
+
+    /**
+     * Run the DAO command on the database connection
+     * @param c
+     * @return
+     * @throws FHIRPersistenceException
+     */
+    public List<ResourceChangeLogRecord> run(Connection c) throws FHIRPersistenceException {
+        ArrayList<ResourceChangeLogRecord> result = new ArrayList<>();
+
+        StringBuilder query = new StringBuilder();
+        query.append(" SELECT c.resource_id, rt.resource_type, lr.logical_id, c.change_tstamp, c.version_id, c.change_type")
+            .append("  FROM ")
+            .append(schemaName).append(".resource_change_log  c, ")
+            .append(schemaName).append(".logical_resources   lr, ")
+            .append(schemaName).append(".resource_types      rt  ")
+            .append(" WHERE lr.logical_resource_id = c.logical_resource_id ")
+            .append("   AND rt.resource_type_id = c.resource_type_id ")
+            ;
+
+        if (fromTstamp != null) {
+            query.append(" AND c.change_tstamp >= ? ");
+        }
+
+        if (afterResourceId != null) {
+            query.append(" AND c.resource_id > ? ");
+        }
+
+        if (resourceTypeName != null) {
+            query.append(" AND rt.resource_type = ? ");
+        }
+
+        query
+            .append(" ORDER BY c.change_tstamp, c.resource_id ")
+            .append(translator.limit(Integer.toString(resourceCount)))
+            ;
+
+        final String SQL = query.toString();
+
+        if (logger.isLoggable(Level.FINE)) {
+            logger.fine("FETCH CHANGES: " + SQL + "; [" + fromTstamp + ", " + afterResourceId + ", " + resourceTypeName + "]");
+        }
+
+        try (PreparedStatement ps = c.prepareStatement(SQL)) {
+            int a = 1;
+            if (this.fromTstamp != null) {
+                ps.setTimestamp(a++, Timestamp.from(this.fromTstamp), UTC_CALENDAR);
+            }
+
+            if (this.afterResourceId != null) {
+                ps.setLong(a++, this.afterResourceId);
+            }
+
+            if (this.resourceTypeName != null) {
+                ps.setString(a++, this.resourceTypeName);
+            }
+
+            ResultSet rs = ps.executeQuery();
+            while (rs.next()) {
+                ResourceChangeLogRecord.ChangeType ct;
+                switch (rs.getString(6)) {
+                case "C": ct = ChangeType.CREATE; break;
+                case "U": ct = ChangeType.UPDATE; break;
+                case "D": ct = ChangeType.DELETE; break;
```

Comment:
Are there any other operations that we would care about? for instance reindex? I realize that doesn't change the resource, rather the values table.

Reply:
Right, reindex doesn't change the resource and this change log is closely tied to that, along with the history operation. We may want to consider an admin log feature, which would track admin operations (like the bulk ops, reindex etc).